src/Components/Web.JS/src/Services/NavigationEnhancement.ts

```diff
@@ -140,6 +140,7 @@ function onDocumentSubmit(event: SubmitEvent) {
     }
 
     performEnhancedPageLoad(url.toString(), /* interceptedLink */ false, fetchOptions);
+    formElem.reset();
```

Comment:
This is very different from the kind of solution I was expecting, and TBH am not sure if it's valid. Doesn't this cause the form to instantly revert to its default state? If so that would be a bug, as it would flash back to being blank while the async form submission is in process. And if the POST request fails, the entire form state would be lost.

The approach I was expecting was to add another case to the DOM sync logic in https://github.com/dotnet/aspnetcore/blob/main/src/Components/Web.JS/src/Rendering/DomMerging/DomSync.ts. There's already a special case for checkboxes at https://github.com/dotnet/aspnetcore/blob/main/src/Components/Web.JS/src/Rendering/DomMerging/DomSync.ts#L351, so my guess (which I haven't verified and could be wrong) is that a similar special case is needed for radio buttons.

There's a unit test for checkboxes at https://github.com/dotnet/aspnetcore/blob/main/src/Components/Web.JS/test/DomSync.test.ts#L472, so hopefully it would be possible to create an equivalent test for radio buttons, showing that it fails with the existing implementation but passes after updating the `DomSync.ts` logic.

This approach avoids resetting the form completely and hence won't discard the user's data from the DOM.

Reply:
Thank you @SteveSandersonMS !
I investigated further the files you mentioned and found out that the same fix as for checkbox won't work for radio button.
I then tracked where exactly checkbox gets updated value on checked attribute https://github.com/dotnet/aspnetcore/blob/main/src/Components/Web.JS/src/Rendering/DomMerging/DomSync.ts#L371 and applied same thing for radio button. 
After this change the current bug was fixed, however the bug you mentioned appeared. https://github.com/dotnet/aspnetcore/blob/main/src/Components/Web.JS/test/DomSync.test.ts#L472 similar test for radio button was failing. I applied the fix you suggested and everything worked.
Since the logic for resetting checkbox and radio button is the same, I added checkbox to my e2e for enhanced form too.